cmd/kubeadm/app/cmd/upgrade/common.go

```diff
@@ -223,6 +224,16 @@ func enforceRequirements(flags *applyPlanFlags, args []string, dryRun bool, upgr
 		printConfiguration(&cfg.ClusterConfiguration, os.Stdout, printer)
 	}
 
+	// Handle a dupliate prefix URL scheme in the Node CRI socket.
+	// Older versions of kubeadm(v1.24.0~2) upgrade may add one or two extra prefix `unix://`
+	// TODO: this fix can be removed in 1.26 once all user node sockets have a correct URL scheme:
+	// https://github.com/kubernetes/kubeadm/issues/2426
+	dupURLScheme := strings.HasPrefix(cfg.NodeRegistration.CRISocket, kubeadmapiv1.DefaultContainerRuntimeURLScheme+"://"+kubeadmapiv1.DefaultContainerRuntimeURLScheme+"://")
+	if dupURLScheme {
+		socket := strings.ReplaceAll(cfg.NodeRegistration.CRISocket, kubeadmapiv1.DefaultContainerRuntimeURLScheme+"://", "")
+		cfg.NodeRegistration.CRISocket = kubeadmapiv1.DefaultContainerRuntimeURLScheme + "://" + socket
```

Comment:
the logic in  `runKubeletConfigPhase()` for `upgrade node` will write the fixed cri socket to remove dups.
but during `apply` are we going to update it as well?
this `cfg` here might be only for local use, but i could be wrong.

Reply:
According to my test, this will be called during `kubeadm config image pull`. To fix it during `kubeadm config image pull` would be werid, but it should be fixed.

- The `upgrade node` will be fixed by the `runKubeletConfigPhase()`.
- `upgrade apply` will call `PerformPostUpgradeTasks` that will call `AnnotateCRISocket` to overwrite the annotation.

Thats why this can fix both.

Reply:
> According to my test, this will be called during `kubeadm config image pull`. To fix it during `kubeadm config image pull` would be werid, but it should be fixed.

`this` here means the common `enforceRequirements`.

Reply:
> According to my test, this will be called during kubeadm config image pull

hm, we should not mutate the CRI socket during actions that are not upgrade.

basically if the socket is corrupted (too many prefixes) upgrade should fix it early in the upgrade phase (enforceRequirements is a good location it seems)

Reply:
Updated.

BTW
- directly run `kubeadm config image pull` can success with dup `unix://` prefix annotation on the node.
- but `kubeadm upgrade apply v1.xx` will fail to run `kubeadm config image pull`
- - with this fix, we will update the annotation in `enforceRequirements`. This can be fixed.